### PR TITLE
Introduce AutoManeuverSystem for BVLOS applications

### PR DESCRIPTION
**Pull request by bresch**
## Description
A basic requirement for BVLOS applications is to allow the drone to perform some emergency maneuvers to prevent collision with manned vehicles or to quickly land in case of low battery/engine failure while flying at high altitudes.

The proposed solution is a new navigator/commander mode, similar to RTL, called Auto Maneuver System (AMS) that can be triggered by the operator or by a companion computer sensing the surrounding environment.

The ways to engage the AMS are: 

*  RC switch
* MAVLink command mode change

Depending on the type of maneuver pre-configured by `AMS_TYPE`, the drone will perform an emergency descent at a `AMS_DESCEND_VEL` speed until `AMS_DESCEND_ALT` above home altitude (or above ground in a rangefinder is used) and then loiter there (`AMS_TYPE` = 0) or land (`AMS_TYPE` = 1).

### Design choices
Since this mode is most likely triggered at cruise speed, a xy velocity waypoint is created with the desired altitude under the drone. Using a velocity waypoint ensures smooth xy deceleration while descending.

In order to go down faster than `MPC_Z_VEL_MAX_DN` currently used to define the desired down velocity in addition to saturate the velocity controller, a new parameter `MPC_Z_VEL_DN` has been introduced. The new meaning is then:

* `MPC_Z_VEL_DN` defines the maximum desired down velocity in manual position control and auto modes
* `MPC_Z_VEL_MAX_DN` defines the maximum down velocity that the drone can safely reach and is used to saturate the velocity controller
With the constraints `MPC_Z_VEL_DN` < or = `AMS_DESCEND_VEL` < `MPC_Z_VEL_MAX_DN`

### AMS triggers
**RC switch**
Set `RC_MAP_AMS_SW` to the desired channel and adjust `RC_AMS_TH`

**Dronecode SDK**
See https://github.com/Dronecode/DronecodeSDK/pull/676

**MAVLink message**
MAV_CMD_DO_SET_MODE (https://mavlink.io/en/messages/common.html#MAV_CMD_DO_SET_MODE)
param1 (base_mode) = 1 (VEHICLE_MODE_FLAG_CUSTOM_MODE_ENABLED)
param2 (custom_main_mode) = 4 (PX4_CUSTOM_MAIN_MODE_AUTO)
param3 (custom_sub_mode) = 10 (PX4_CUSTOM_SUB_MODE_AUTO_AMS)

**Shell**
`nsh>commander mode auto:ams`

### CODE REVIEW

**Comment by dagar** [comment]
Awesome! One thing that's become a little unclear is the line between navigator states and flight tasks.

File: src/modules/navigator/navigation.h [hunk 1]
```diff
@@ -142,11 +143,20 @@ struct mission_item_s {
...
-			float ___lat_float;			/**< padding */
-			float ___lon_float;			/**< padding */
+			float acceptance_radius;	/**< default radius in which the mission is accepted as reached in meters */
+			union {
+				struct {
+					float loiter_radius;		/**< loiter radius in meters, 0 for a VTOL to hover, negative for counter-clockwise */
+					float ___lat_float;			/**< padding */
+					float ___lon_float;			/**< padding */
+				};
+				struct {
```
Comment (dagar) [inline src/modules/navigator/navigation.h]: I need to dig into AMS a bit more, but you likely don't need to extend `mission_item_s` like this unless you're checking for velocity setup acceptance (`MissionBlock::is_mission_item_reached()`).

File: src/modules/commander/state_machine_helper.cpp [hunk 2]
```diff
@@ -314,6 +314,7 @@ main_state_transition(const vehicle_status_s &status, const main_state_t new_mai
 
 		break;
 
+	case commander_state_s::MAIN_STATE_AUTO_AMS:
```
Comment (dagar) [inline src/modules/commander/state_machine_helper.cpp]: I would also limit this to multicopter.

File: src/modules/navigator/navigation.h (hunk 1 above)
Comment (bresch) [reply src/modules/navigator/navigation.h]: How should I do then? I wanted to add vx, vy and vz fields without increasing the size of mission_item_s. That's why I selected three floats I don't need for the velocity waypoint to put them in the same union.

File: src/modules/commander/state_machine_helper.cpp (hunk 2 above)
Comment (bresch) [reply src/modules/commander/state_machine_helper.cpp]: Done

**Comment by bresch** [comment]
> One thing that's become a little unclear is the line between navigator states and flight tasks.

@dagar In that case, I think the line is quite clear: the navigator logic provides the desired position/velocity waypoints and the Auto FlightTask executes it as regular waypoints. It's like for RTL, the FlightTask doesn't know and should not know that we are performing an AMS/RTL. 
In my opinion, the FlightTasks should not handle too much high level logic; we don't want it to become a new Commander/Navigator...

**Comment by davids5** [comment]
devcall AI:@bresch will rebase @dagar to review

**Comment by bresch** [comment]
@dagar I rebased it onto master

**Comment by thomasgubler** [comment]
@bresch could you rebase?

@bkueng could you do the review and merge (for post 1.9)

**Comment by bresch** [comment]
Will work on that again once I have the "motion primitives" done.

**Comment by LorenzMeier** [comment]
@bresch Where is this standing?

**Comment by bresch** [comment]
Still waiting for the motion primitives. I close this PR as the implementation will change completely

**Pull request closed without merging**